src/platforms/macos/daemon/dnsutilsmacos.cpp

```diff
@@ -150,41 +156,36 @@ bool DnsUtilsMacos::updateResolvers(const QString& ifname,
     SCDynamicStoreSetValue(m_scStore, dnsPath, dnsConfig);
     CFRelease(dnsPath);
   }
+
   return true;
 }
 
 bool DnsUtilsMacos::restoreResolvers() {
-  for (const QString& uuid : m_prevServices.keys()) {
-    CFStringRef path = CFStringCreateWithFormat(
-        kCFAllocatorSystemDefault, nullptr,
-        CFSTR("Setup:/Network/Service/%s/DNS"), qPrintable(uuid));
-
-    logger.debug() << "Restoring DNS config for" << uuid;
-    const DnsBackup& backup = m_prevServices[uuid];
-    if (backup.isValid()) {
-      CFMutableDictionaryRef config;
-      config = CFDictionaryCreateMutable(kCFAllocatorSystemDefault, 0,
-                                         &kCFCopyStringDictionaryKeyCallBacks,
-                                         &kCFTypeDictionaryValueCallBacks);
+  if (m_dnsSnapshotPid < 0) {
+    logger.debug() << "No DNS resolvers to restore";
+    return false;
+  }
 
-      cfDictSetString(config, kSCPropNetDNSDomainName, backup.m_domain);
-      cfDictSetStringList(config, kSCPropNetDNSSearchDomains, backup.m_search);
-      cfDictSetStringList(config, kSCPropNetDNSServerAddresses,
-                          backup.m_servers);
-      cfDictSetStringList(config, kSCPropNetDNSSortList, backup.m_sortlist);
-      SCDynamicStoreSetValue(m_scStore, path, config);
-      CFRelease(config);
-    } else {
-      SCDynamicStoreRemoveValue(m_scStore, path);
-    }
-    CFRelease(path);
+  // Kill the DNS watchdog to restore the resolver configuration.
+  auto guard = qScopeGuard([&] { m_dnsSnapshotPid = -1; });
+  if (kill(m_dnsSnapshotPid, SIGINT) < 0) {
+    logger.debug() << "Failed to signal DNS restorer";
+    return false;
   }
 
-  m_prevServices.clear();
+  // Wait for the process to exit, and gather it's status.
+  int status = 0;
+  if (waitpid(m_dnsSnapshotPid, &status, 0) < 0) {
```

Comment:
To protect against a hung child process, should WNOHANG be used?

Reply:
This should now be addressed by using a `QProcess` class, which waits for at most 30s when trying to terminate a child process.